crates/libp2p-networking/src/network/stake_table_transport.rs

```diff
@@ -0,0 +1,749 @@
+use anyhow::Result as AnyhowResult;
+use anyhow::{ensure, Context};
+use async_compatibility_layer::art::async_timeout;
+use futures::AsyncRead;
+use futures::AsyncWrite;
+use serde::Deserialize;
+use serde::Serialize;
+use std::collections::HashSet;
+use std::future::Future;
+use std::hash::BuildHasher;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::Poll;
+use tracing::warn;
+
+use futures::future::poll_fn;
+use futures::{AsyncReadExt, AsyncWriteExt};
+use hotshot_types::traits::signature_key::SignatureKey;
+use libp2p::core::muxing::StreamMuxerExt;
+use libp2p::core::transport::TransportEvent;
+use libp2p::core::StreamMuxer;
+use libp2p::identity::PeerId;
+use libp2p::Transport;
+use pin_project::pin_project;
+
+/// The maximum size of an authentication message. This is used to prevent
+/// DoS attacks by sending large messages.
+const MAX_AUTH_MESSAGE_SIZE: usize = 1024;
+
+/// The timeout for the authentication handshake. This is used to prevent
+/// attacks that keep connections open indefinitely by half-finishing the
+/// handshake.
+const AUTH_HANDSHAKE_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(5);
+
+/// A wrapper for a `Transport` that bidirectionally authenticates connections
+/// by performing a handshake that checks if the remote peer is present in the
+/// stake table.
+#[pin_project]
+pub struct StakeTableAuthentication<T: Transport, S: SignatureKey, C: StreamMuxer + Unpin> {
+    #[pin]
+    /// The underlying transport we are wrapping
+    pub inner: T,
+
+    /// The stake table we check against to authenticate connections
+    pub stake_table: Arc<Option<HashSet<S>>>,
+
+    /// A pre-signed message that we send to the remote peer for authentication
+    pub auth_message: Arc<Option<Vec<u8>>>,
+
```

Comment:
Is it Optional because of upgrading, or some other reason?

Reply:
Yeah it is for upgrading, but I am open to making it non-optional

Reply:
Long run non optional what we want, but for now it makes sense